GridServices/Control/ILCAgent/ilc/utils.py

```
@@ -42,64 +42,123 @@
     rep = dict((re.escape(k), v) for k, v in rep.items())
     pattern = re.compile("|".join(rep.keys()))
     new_key = pattern.sub(lambda m: rep[re.escape(m.group(0))], text)
     return new_key
 
 
-def parse_sympy(data, condition=False):
+def sympy_evaluate(condition: str, points: Iterable[Tuple[str, float]]) -> bool | float:
     """
-    :param condition:
-    :param data:
-    :return:
+    Calls clean_text to remove special characters from string in points,
+    does string replace to for cleaned point in condition, and evaluates symbolic math
+    condition.
+
+    :param condition: string equation or condition.
+    :type condition: str
+    :param points: list of tuples with - [(point_name, value)] =
+    :type points: list[tuples]
+    :return: evaluated sympy expression
+    :rtype: float or bool
     """
-    if isinstance(data, dict):
-        return_data = {}
-        for key, value in data.items():
-            new_key = clean_text(key)
-            return_data[new_key] = value
-
-    elif isinstance(data, list):
-        if condition:
-            return_data = ""
-            for item in data:
-                parsed_string = clean_text(item)
-                parsed_string = "(" + clean_text(item) + ")" if parsed_string not in ("&", "|") else parsed_string
-                return_data += parsed_string
-        else:
-            return_data = []
-            for item in data:
-                return_data.append(clean_text(item))
+    cleaned_points = []
+    cleaned_condition = condition
+    for point, value in points:
+        cleaned = clean_text(point)
+        cleaned_condition = cleaned_condition.replace(point, cleaned)
+        cleaned_points.append((cleaned, value))
+    _log.debug(f"Sympy debug condition: {condition} -- {cleaned_condition}")
+    _log.debug(f"Sympy debug points: {points} -- {cleaned_points}")
+    equation = parse_expr(cleaned_condition)
+    return_value = equation.subs(cleaned_points)
```

What if there ValueError at this point? There is no catch statement/error handling.

---

It is probably preferable for it to break or throw an exception rather than continue on in vain.